### PR TITLE
Add listing type and new categories elements to offer form

### PR DESCRIPTION
**Pull request by SamTremko**

### CODE REVIEW

**Comment by SamTremko** [comment]
* **#910** <a href="https://app.graphite.dev/github/pr/vexl-it/vexl/910?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#905** <a href="https://app.graphite.dev/github/pr/vexl-it/vexl/905?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @SamTremko and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/resources-utils/src/offers/utils/encryptOfferPublicPayload.ts [hunk 1]
```diff
@@ -50,6 +50,9 @@ const OfferPublicPartToEncrypt = z.object({
   activePriceCurrency: z.string(),
   active: z.enum(['true', 'false']),
   groupUuids: z.array(z.string()),
+  listingType: z.string(),
```
Comment (kaladivo) [inline packages/resources-utils/src/offers/utils/encryptOfferPublicPayload.ts:53]: Map to locaitonState

File: packages/domain/src/general/offers.ts [hunk 2]
```diff
@@ -146,6 +152,9 @@ export const OfferPublicPart = z.object({
   activePriceCurrency: CurrencyCode,
   active: z.boolean(),
   groupUuids: z.array(z.string()),
+  listingType: ListingType.optional(),
+  singlePrice: z.number().optional(),
+  deliveryMethod: z.array(DeliveryMethod).default([]),
```
Comment (kaladivo) [inline packages/domain/src/general/offers.ts:157]: Use location state with fallback for older versions (in resource-utils)

File: packages/resources-utils/src/offers/utils/encryptOfferPublicPayload.ts [hunk 3]
```diff
@@ -50,6 +50,9 @@ const OfferPublicPartToEncrypt = z.object({
...
+  deliveryMethod: z.array(z.string()).optional(),
```
Comment (kaladivo) [inline packages/resources-utils/src/offers/utils/encryptOfferPublicPayload.ts]: Use location state with fallback for older versions (in resource-utils)

File: packages/domain/src/general/offers.ts (hunk 2 above)
Comment (kaladivo) [inline packages/domain/src/general/offers.ts]: Use location state with fallback for older versions (in resource-utils)

File: packages/domain/src/general/offers.ts [hunk 4]
```diff
@@ -146,6 +152,9 @@ export const OfferPublicPart = z.object({
   activePriceCurrency: CurrencyCode,
   active: z.boolean(),
   groupUuids: z.array(z.string()),
+  listingType: ListingType.optional(),
+  singlePrice: z.number().optional(),
```
Comment (kaladivo) [inline packages/domain/src/general/offers.ts]: use amountBottomLimit/amountTopLimit

File: packages/domain/src/general/offers.ts [hunk 5]
```diff
@@ -134,7 +140,7 @@ export const OfferPublicPart = z.object({
   amountTopLimit: z.coerce.number(),
   feeState: FeeState,
   feeAmount: z.coerce.number(),
-  locationState: LocationState,
+  locationState: LocationState.default('IN_PERSON'),
```
Comment (kaladivo) [inline packages/domain/src/general/offers.ts]: No default, always set when creating offer

File: apps/mobile/src/components/CurrentBtcPrice.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import {Text, XStack, getTokens, type TextProps} from 'tamagui'
+import {getCurrentLocale} from '../utils/localization/I18nProvider'
+import {currencies} from '../utils/localization/currency'
+import {preferencesAtom} from '../utils/preferences'
+
+interface Props extends TextProps {
+  customBtcPrice?: number
+  refreshBtcPrice: () => void
+  currencyAtom: PrimitiveAtom<CurrencyCode | undefined>
+  btcPriceWithStateAtom: Atom<BtcPriceDataWithState | undefined>
```
Comment (kaladivo) [inline apps/mobile/src/components/CurrentBtcPrice.tsx]: Make this atom accept only currency code and interact with currentBtcPriceAtom directly

File: apps/mobile/src/components/InsideRouter/components/BitcoinPriceChart.tsx [hunk 7]
```diff
@@ -23,8 +23,8 @@ function BitcoinPriceChart(): JSX.Element {
 
   useFocusEffect(
     useCallback(() => {
-      void refreshBtcPrice(selectedCurrency)()
-    }, [refreshBtcPrice, selectedCurrency])
+      void refreshBtcPrice(selectedCurrencyAtom)()
```
Comment (kaladivo) [inline apps/mobile/src/components/InsideRouter/components/BitcoinPriceChart.tsx]: Make sure that the price is refreshed when selected currency changes

File: apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts [hunk 8]
```diff
@@ -76,6 +86,112 @@ function getAtomWithNullableValueHandling<T, S>(
...
+    deliveryMethod,
+    offerDescription,
+    ...restOfPublicPart
+  } = offerPublicPart
+  return listingType === 'BITCOIN'
+    ? ({
+        listingType,
+        ...restOfPublicPart,
+        offerDescription: offerDescription.trim(),
+      } as OfferPublicPart)
```
Comment (kaladivo) [inline apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts]:
```suggestion
      } satisfies OfferPublicPart)
```

File: apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts [hunk 9]
```diff
@@ -76,6 +86,112 @@ function getAtomWithNullableValueHandling<T, S>(
...
+    : listingType === 'PRODUCT'
+    ? ({
+        listingType,
+        singlePriceState,
+        singlePriceValue,
+        deliveryMethod,
+        ...restOfPublicPart,
+        offerDescription: offerDescription.trim(),
+      } as OfferPublicPart)
```
Comment (kaladivo) [inline apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts]:
```suggestion
      } satisfies OfferPublicPart)
```

File: apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts [hunk 10]
```diff
@@ -76,6 +86,112 @@ function getAtomWithNullableValueHandling<T, S>(
...
+    : ({
+        listingType,
+        singlePriceState,
+        singlePriceValue,
+        ...restOfPublicPart,
+        offerDescription: offerDescription.trim(),
+      } as OfferPublicPart)
```
Comment (kaladivo) [inline apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts]:
```suggestion
      } satisfies OfferPublicPart)
```

File: apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts [hunk 11]
```diff
@@ -381,53 +526,39 @@ export const offerFormMolecule = molecule(() => {
...
-      ...restOfPublicPart
-    } = get(offerFormAtom)
-
-    const intendedConnectionLevel = get(intendedConnectionLevelAtom)
-    const belowProgresLeft = get(modifyOfferLoaderTitleAtom)
-
-    if (locationState === 'IN_PERSON' && location.length === 0) {
-      Alert.alert(t('offerForm.errorLocationNotFilled'))
+    if (
+      !checkConditionsToCreateOfferAreMet({offerForm: get(offerFormAtom), t})
```
Comment (kaladivo) [inline apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts]: Rename to `checkConditionsToCreateOfferAreMetAlertIfNot`

File: apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts [hunk 12]
```diff
@@ -720,9 +853,90 @@ export const offerFormMolecule = molecule(() => {
...
+      if (currentBtcPrice) {
+        set(
+          singlePriceValueAtom,
+          Math.round(currentBtcPrice * (satsNumber / SATOSHIS_IN_BTC))
+        )
+      }
+    }
+  )
+
+  const toggleCurrencyActionAtom = atom(
```
Comment (kaladivo) [inline apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts]: setCurrencyActionAtom

File: apps/mobile/src/components/ModifyOffer/useBtcOfferContent.tsx [hunk 13]
```diff
@@ -75,21 +77,21 @@ export default function useContent(): Props[] {
         ),
       },
       {
-        title: t('offerForm.premiumOrDiscount.premiumOrDiscount'),
         customSection: true,
         children: (
           <PremiumOrDiscount
+            key={t('offerForm.premiumOrDiscount.premiumOrDiscount')}
```
Comment (kaladivo) [inline apps/mobile/src/components/ModifyOffer/useBtcOfferContent.tsx]: Should the key be there?

File: apps/mobile/src/components/ModifyOffer/useBtcOfferContent.tsx [hunk 14]
```diff
@@ -75,21 +77,21 @@ export default function useContent(): Props[] {
...
             feeStateAtom={feeStateAtom}
           />
         ),
       },
       {
-        title: t('offerForm.expiration.expiration'),
         customSection: true,
         children: (
           <Expiration
+            key={t('offerForm.expiration.expiration')}
```
Comment (kaladivo) [inline apps/mobile/src/components/ModifyOffer/useBtcOfferContent.tsx]: Should the key be there?

File: apps/mobile/src/components/ModifyOffer/useProductOfferContent.tsx [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+            offerTypeAtom={offerTypeAtom}
+          />
+        ),
+        mandatory: true,
+      },
+      {
+        customSection: true,
+        children: (
+          <Price
+            key={t('offerForm.price')}
```
Comment (kaladivo) [inline apps/mobile/src/components/ModifyOffer/useProductOfferContent.tsx]: Check keys in this file also

File: apps/mobile/src/components/OfferForm/components/ListingType/useContent.tsx [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+import {type TabProps} from '../../../Tabs'
+
+export default function useContent(): Array<TabProps<ListingType>> {
+  const {t} = useTranslation()
+
+  return useMemo(
+    () => [
+      {
+        title: t('offerForm.BITCOIN'),
+        type: 'BITCOIN' as ListingType,
```
Comment (kaladivo) [inline apps/mobile/src/components/OfferForm/components/ListingType/useContent.tsx]: use `as const`

File: apps/mobile/src/components/OfferForm/index.tsx [hunk 17]
```diff
@@ -18,18 +18,10 @@ function OfferForm({content}: {content: Props[]}): JSX.Element {
     <ChunkView displayOnProgress={<LoaderComponent />}>
       {content.map((item) =>
         item.customSection ? (
```
Comment (kaladivo) [inline apps/mobile/src/components/OfferForm/index.tsx:20]: `<Fragment key={...}>`

<script>alert('ajaj');</script>

File: apps/mobile/src/state/marketplace/atoms/offersToSeeInMarketplace.ts [hunk 18]
```diff
@@ -1,38 +1,40 @@
...
 export function alertAndReportOnlineOffersWithoutLocation(
   offers: OfferInfo[],
   reportOffersPayload: boolean = false,
   showAlert: boolean = false
 ): void {
-  const onlineOffersWithoutLocation = offers.filter(
+  const inPersonOffersWithoutLocation = offers.filter(
     (one) =>
-      one.publicPart.locationState === 'IN_PERSON' &&
+      one.publicPart.listingType !== 'OTHER' &&
```
Comment (kaladivo) [inline apps/mobile/src/state/marketplace/atoms/offersToSeeInMarketplace.ts:16]: Make sure other offer has always locationState set

**Review by kaladivo — CHANGES_REQUESTED** [review]

File: apps/mobile/src/utils/bigNumberToString.ts [hunk 19]
```diff
@@ -3,5 +3,9 @@ export function bigNumberToString(number: number): string {
     return `${Math.round(number / 1_000)}K`
   }
 
+  if (number > 1_000_000) {
+    return `${Math.round(number / 1_000)}M`
```
Comment (SamTremko) [inline apps/mobile/src/utils/bigNumberToString.ts]: check this conversion

File: apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts [hunk 20]
```diff
@@ -133,6 +253,7 @@ export const dummyOffer: OneOfferInState = {
 
 export const offerFormMolecule = molecule(() => {
   const offerAtom = atom<OneOfferInState>(dummyOffer)
+  const nullableListingTypeAtom = atom<ListingType | undefined>(undefined)
   const nullableOfferTypeAtom = atom<OfferType | undefined>(undefined)
```
Comment (SamTremko) [inline apps/mobile/src/components/ModifyOffer/atoms/offerFormStateAtoms.ts:243]: change offer type for product to oposite compared to BTC offer

File: apps/mobile/src/components/OfferForm/components/LocationsList.tsx [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+function LocationsList({
+  locations,
+  onLocationRemove,
+}: Props): JSX.Element[] | null {
+  const {t} = useTranslation()
+
+  return locations
+    ? locations?.map((loc) => (
+        <XStack
+          key={loc.latitude}
```
Comment (kaladivo) [inline apps/mobile/src/components/OfferForm/components/LocationsList.tsx:26]: loc has Id. Use id as a key

Comment (kaladivo) [inline apps/mobile/src/components/OfferForm/components/LocationsList.tsx]:
```suggestion
  key={loc.placeId}
```

Comment (kaladivo) [inline apps/mobile/src/components/OfferForm/components/LocationsList.tsx:26]:
```suggestion
  key={loc.placeId}
```

File: packages/resources-utils/src/offers/utils/encryptOfferPublicPayload.ts
Comment (kaladivo) [inline packages/resources-utils/src/offers/utils/encryptOfferPublicPayload.ts:1]: Remove v2 where it does not make sense

**Review by kaladivo — APPROVED** [review]

**Comment by SamTremko** [comment]
### Merge activity

* **Mar 25, 10:08 AM EDT**: @SamTremko merged this pull request with [Graphite](https://app.graphite.dev/github/pr/vexl-it/vexl/905).

**Pull request merged**